File: src/main/java/io/myosotisdev/utopianism/modules/stat/ItemStat.java
package io.myosotisdev.utopianism.modules.stat;

import io.myosotisdev.utopianism.battle.DamageMeta;
import io.myosotisdev.utopianism.modules.stat.data.ItemStatData;
import io.myosotisdev.utopianism.modules.stat.type.BooleanStat;
import io.myosotisdev.utopianism.modules.stat.type.DoubleStat;
import io.myosotisdev.utopianism.modules.stat.type.IntegerStat;
import io.myosotisdev.utopianism.modules.stat.type.StringStat;
import net.minestom.server.entity.EntityProjectile;
import net.minestom.server.item.ItemStack;
import net.minestom.server.tag.Tag;

public abstract class ItemStat<T>
{
    public static final ItemStat ITEM_ID = new StringStat("id");
    public static final ItemStat ITEM_SLOT = new EquipSlot();
    public static final ItemStat DISABLE_INTERACTION = new BooleanStat("disable-interaction");
    public static final ItemStat DISABLE_ARROW_SHOOTING = new BooleanStat("disable-arrow-shooting");
    public static final ItemStat REQUIRED_CLASS_LEVEL = new RequiredLevel();
    public static final ItemStat REQUIRED_CAREER_LEVEL = new RequiredLevel();
    public static final ItemStat ATTACK_DAMAGE = new AttackDamage();
    public static final ItemStat ATTACK_SPEED = new AttackSpeed();
    public static final ItemStat ATTACK_RANGE = new AttackRange();
    public static final ItemStat CRITICAL_STRIKE_CHANCE = new DoubleStat("critical-strike-chance");
    public static final ItemStat CRITICAL_STRIKE_POWER = new DoubleStat("critical-strike-power");
    public static final ItemStat BLOCK_CHANCE = new DoubleStat("block-chance");
    public static final ItemStat BLOCK_POWER = new DoubleStat("block-power");
    public static final ItemStat DODGE_CHANCE = new DoubleStat("dodge-chance");
    public static final ItemStat DODGE_POWER = new DoubleStat("dodge-power");
    public static final ItemStat MAX_HEALTH = new DoubleStat("max-health");
    public static final ItemStat MAX_MANA = new DoubleStat("max-mana");
    public static final ItemStat HEALTH_REGEN = new HealthRegeneration();
    public static final ItemStat MANA_REGEN = new ManaRegeneration();
    public static final ItemStat ARROW_VELOCITY = new ArrowVelocity();
    public static final ItemStat DAMAGE_RESIST = new DoubleStat("damage-resist");
    public static final ItemStat DAMAGE_RESIST_RATE = new DoubleStat("damage-resist-%");
    public static final ItemStat MAGIC_RESIST = new DoubleStat("magic-resist");
    public static final ItemStat MAGIC_RESIST_RATE = new DoubleStat("magic-resist-%");
    public static final ItemStat UNBREAKABLE = new Unbreakable();
    public static final ItemStat FACTION_DAMAGE = new FactionDamage();
    public static final ItemStat KNOCKBACK = new DoubleStat("knockback");
    public static final ItemStat KNOCKBACK_RESIST = new DoubleStat("knockback-resist");
    public static final ItemStat TWO_HANDED = new BooleanStat("two-handed");
    public static final ItemStat SOULBOUND = new BooleanStat("soulbound");
    public static final ItemStat SOULBOUND_CHANCE = new DoubleStat("soulbound-chance");
    public static final ItemStat REINFORCE = new IntegerStat("reinforce");

    private String key;
    private Tag<T> tag;

    public ItemStat(String key, Tag<T> newTag)
    {
        this.key = key;
        this.tag = newTag;
    }

    public String getKey()
    {
        return key;
    }

    public Tag<T> getTag()
    {
        return tag;
    }

    /**
     * This function is called when generating new item stack
     * and it should apply attributes into the item
     *
     * @param itemStack
     * @param data
     */
    public abstract void onApply(ItemStack itemStack, ItemStatData<T> data);

    public abstract ItemStatData<T> defaultStatData();

    public T getLoadedData(ItemStack itemStack)
    {
        return itemStack.getTag(tag);
    }
}

File: src/main/java/io/myosotisdev/utopianism/modules/stat/ItemStat.kt
package io.myosotisdev.utopianism.modules.stat

import com.google.gson.JsonObject
import io.myosotisdev.utopianism.modules.stat.data.ItemStatData
import io.myosotisdev.utopianism.modules.stat.type.BooleanStat
import io.myosotisdev.utopianism.modules.stat.type.DoubleStat
import io.myosotisdev.utopianism.modules.stat.type.IntegerStat
import io.myosotisdev.utopianism.modules.stat.type.StringStat
import net.minestom.server.item.ItemStack
import net.minestom.server.tag.Tag

abstract class ItemStat<T>(val key: String, val tag: Tag<T>)
{

    init { stats[key] = this }

    /**
     * This function is called when generating new item stack
     * and it should apply attributes into the item
     *
     * @param itemStack
     * @param data
     */
    open fun onApply(itemStack: ItemStack, data: ItemStatData<out T>)
    {
        itemStack.withTag(tag, data.value)
    }

    abstract fun defaultStatData(): ItemStatData<T>?

    abstract fun createStatData(jsonObject: JsonObject): ItemStatData<T>

    open fun getLoadedData(itemStack: ItemStack): T
    {
        return itemStack.getTag(tag)
    }

    companion object
    {
        private val stats: HashMap<String, ItemStat<*>> = HashMap()

        val ITEM_ID: ItemStat<*> = StringStat("id")
        val ITEM_SLOT: ItemStat<*> = EquipSlot()
        val DISABLE_INTERACTION: ItemStat<*> = BooleanStat("disable-interaction")
        val DISABLE_ARROW_SHOOTING: ItemStat<*> = BooleanStat("disable-arrow-shooting")
        val REQUIRED_CLASS_LEVEL: ItemStat<*> = RequiredLevel()
        val REQUIRED_CAREER_LEVEL: ItemStat<*> = RequiredLevel()
        val ATTACK_DAMAGE: ItemStat<*> = AttackDamage()
        val ATTACK_SPEED: ItemStat<*> = AttackSpeed()
        val ATTACK_RANGE: ItemStat<*> = AttackRange()
        val CRITICAL_STRIKE_CHANCE: ItemStat<*> = DoubleStat("critical-strike-chance")
        val CRITICAL_STRIKE_POWER: ItemStat<*> = DoubleStat("critical-strike-power")
        val BLOCK_CHANCE: ItemStat<*> = DoubleStat("block-chance")
        val BLOCK_POWER: ItemStat<*> = DoubleStat("block-power")
        val DODGE_CHANCE: ItemStat<*> = DoubleStat("dodge-chance")
        val DODGE_POWER: ItemStat<*> = DoubleStat("dodge-power")
        val MAX_HEALTH: ItemStat<*> = DoubleStat("max-health")
        val MAX_MANA: ItemStat<*> = DoubleStat("max-mana")
        val HEALTH_REGEN: ItemStat<*> = HealthRegeneration()
        val MANA_REGEN: ItemStat<*> = ManaRegeneration()
        val ARROW_VELOCITY: ItemStat<*> = ArrowVelocity()
        val DAMAGE_RESIST: ItemStat<*> = DoubleStat("damage-resist")
        val DAMAGE_RESIST_RATE: ItemStat<*> = DoubleStat("damage-resist-%")
        val MAGIC_RESIST: ItemStat<*> = DoubleStat("magic-resist")
        val MAGIC_RESIST_RATE: ItemStat<*> = DoubleStat("magic-resist-%")
        val UNBREAKABLE: ItemStat<*> = Unbreakable()
        val FACTION_DAMAGE: ItemStat<*> = FactionDamage()
        val KNOCKBACK: ItemStat<*> = DoubleStat("knockback")
        val KNOCKBACK_RESIST: ItemStat<*> = DoubleStat("knockback-resist")
        val TWO_HANDED: ItemStat<*> = BooleanStat("two-handed")
        val SOULBOUND: ItemStat<*> = BooleanStat("soulbound")
        val SOULBOUND_CHANCE: ItemStat<*> = DoubleStat("soulbound-chance")
        val REINFORCE: ItemStat<*> = IntegerStat("reinforce")

        fun getByName(key: String): ItemStat<*>? = stats[key]
    }
}